src/debug/debugController.ts

```diff
@@ -25,23 +25,27 @@ export function installDebugController() {
 }
 
 class DebugController implements ContextListener {
-  private async ensureInstalledInFrame(frame: frames.Frame) {
-    try {
-      await frame.extendInjectedScript(debugScriptSource.source);
-    } catch (e) {
-    }
-  }
-
   async onContextCreated(context: BrowserContext): Promise<void> {
-    if (!isDebugMode())
-      return;
-    context.on(BrowserContext.Events.Page, (page: Page) => {
-      for (const frame of page.frames())
-        this.ensureInstalledInFrame(frame);
-      page.on(Page.Events.FrameNavigated, frame => this.ensureInstalledInFrame(frame));
-    });
+    if (isDebugMode())
+      installDebugControllerInContext(context);
   }
-
   async onContextWillDestroy(context: BrowserContext): Promise<void> {}
   async onContextDidDestroy(context: BrowserContext): Promise<void> {}
 }
+
+async function ensureInstalledInFrame(frame: frames.Frame) {
+  try {
+    await frame.extendInjectedScript(debugScriptSource.source);
+  } catch (e) {
+  }
+}
+
+async function installInPage(page: Page) {
+  page.on(Page.Events.FrameNavigated, ensureInstalledInFrame);
```

Comment:
> Won't this miss iframes without src?

Good point! I've investigated this, and I am happy to report that all our frames issue `framenavigated` at least once. Added a test just in case.

> And why do this at all instead of evaluateOnNewDocument?

We use `frame.extendInjectedScript` to get access to injected script, which is not available to `evaluateOnNewDocument` scripts.

Reply:
I would expect a pattern of each script adds a callback into window.injectedExtensions. Then when the injected script eventually loads, it reads all extensions. Kind of like webpack.